packages/backend/server/src/core/doc-renderer/controller.ts

```diff
@@ -0,0 +1,112 @@
+import { Controller, Get, Param, Res } from '@nestjs/common';
+import type { Response } from 'express';
+
+import { DocNotFound, URLHelper } from '../../fundamentals';
+import { DocManager } from '../doc';
+import { PermissionService } from '../permission';
+import { PageDocContent, PageDocContentType } from '../utils/blocksuite';
+
+@Controller('/workspace/:workspaceId/:docId')
+export class DocRendererController {
+  constructor(
+    private readonly doc: DocManager,
+    private readonly permission: PermissionService,
+    private readonly url: URLHelper
+  ) {}
+
+  @Get()
+  async render(
+    @Res() res: Response,
+    @Param('workspaceId') workspaceId: string,
+    @Param('docId') docId: string
+  ) {
+    if (workspaceId === docId) {
+      throw new DocNotFound({ workspaceId, docId });
+    }
+
+    // if page is public, show all
+    // if page is private, but workspace public og is on, show og but not content
+    let og = false;
+    let content = false;
+    const isPagePublic = await this.permission.isPublicPage(workspaceId, docId);
+
+    if (isPagePublic) {
+      og = true;
+      content = true;
+    } else {
+      const allowPreview = await this.permission.allowUrlPreview(workspaceId);
+
+      if (allowPreview) {
+        og = true;
+      }
```

Comment:
for private workspace, we can return a page that include <meta name="robots" content="noindex, nofollow"> to prevent the search engine indexing
otherwise, when a private workspace is referenced by a public workspace page, but the search engine get a 404 response by the link, the website rate will be downgraded